Added KPopLyrics

Lukeuke opened the pull request:
Source:
- Provider
- Parser
- UriConverter - custom method to generete special slug formatting for KPopLyrics
- Options

Tests:
- Stantard tests for ExternalProvider (KPopLyricsProvider)
- Unit test for KPopLyricsUriConverter (CreateCombinedUrlSlug)

Lukeuke committed:
Added KPopLyrics

Lukeuke committed:
Whitespace formatting fix

Lukeuke commented:
PR for Issue #39

Lukeuke committed:
Missing KPopLyricsOptions declaration

skuill commented on LyricsScraperNET/Configuration/ILyricScraperClientConfig.cs:
```diff
@@ -24,5 +24,6 @@ public interface ILyricScraperClientConfig
         IExternalProviderOptions SongLyricsOptions { get; }
 
         IExternalProviderOptions LyricFindOptions { get; }
+        IExternalProviderOptions KPopLyricsOptions { get; }
```
Please add an empty line before this for better readability

skuill commented on LyricsScraperNET/Configuration/ServiceCollectionExtensions.cs:
```diff
@@ -8,6 +8,7 @@
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Options;
 using System;
+using LyricsScraperNET.Providers.KPopLyrics;
```
All usings could be sort here

skuill commented on LyricsScraperNET/Providers/KPopLyrics/KPopLyricsUriConverter.cs:
```diff
@@ -0,0 +1,16 @@
+
+        public Uri GetLyricUri(string artist, string song)
+        {
+            return new Uri(_baseUri, $"{StringExtensions.CreateCombinedUrlSlug(artist, song)}-lyrics-english-romanized.html");
```
Suggest to remove `-english-romanized` part. When I checked, it worked without this part. Could you please check? I'm not sure that the lyrics will always work if we add this part to the URI :)

skuill commented on LyricsScraperNET/Providers/KPopLyrics/KPopLyricsProvider.cs:
```diff
@@ -0,0 +1,179 @@
+            var htmlDoc = new HtmlDocument();
+            htmlDoc.LoadHtml(text);
+
+            var mainNode = htmlDoc.DocumentNode.SelectNodes(LyricsContainerNodesXPath).FirstOrDefault();
```
Simplifying node selection logic: Instead of calling `SelectNodes `followed by `FirstOrDefault`, you can use `SelectSingleNode `to avoid additional searching.

skuill commented on LyricsScraperNET/Providers/KPopLyrics/KPopLyricsProvider.cs:
```diff
@@ -0,0 +1,179 @@
+        // we only want the first lyrics
+        private string TakeParagraphsUntilHeader(HtmlNode h2)
+        {
+            var output = string.Empty;
```
Optimizing the `TakeParagraphsUntilHeader `method: Eliminate the need for `string `concatenation by using `StringBuilder`, which will improve performance.

skuill commented on LyricsScraperNET/Providers/KPopLyrics/KPopLyricsUriConverter.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace LyricsScraperNET.Providers.KPopLyrics
+{
+    public class KPopLyricsUriConverter : IExternalUriConverter
```
Suggest to add Unit Tests for `KPopLyricsUriConverter ` similar to other converters. You can refer to test cases in `AZLyricsUriConverterTests` or find something similar on the kpoplyrics website. 
Please ensure that the information is correctly fetched from the specified address when writing the tests.

skuill commented on LyricsScraperNET/Configuration/LyricScraperClientConfig.cs:
```diff
@@ -22,6 +23,7 @@ public sealed class LyricScraperClientConfig : ILyricScraperClientConfig
         public IExternalProviderOptions SongLyricsOptions { get; set; } = new SongLyricsOptions();
 
         public IExternalProviderOptions LyricFindOptions { get; set; } = new LyricFindOptions();
+        public IExternalProviderOptions KPopLyricsOptions { get; set; } = new KPopLyricsOptions();
```
Please add an empty line before this for better readability

skuill commented on LyricsScraperNET/Providers/KPopLyrics/KPopLyricsProvider.cs:
```diff
@@ -0,0 +1,179 @@
+
+            if (mainNode is null)
+            {
+                return new SearchResult(ExternalProviderType.KPopLyrics, ResponseStatusCode.NoDataFound);
```
Suggest adding logging to catch errors early and notice them. We expect the page to have lyrics by this point.
Let's add something like:
` _logger?.LogWarning($"KPopLyrics. Can't parse lyric from the page. Uri: {uri.ToString()}");`

skuill commented on LyricsScraperNET/Providers/KPopLyrics/KPopLyricsProvider.cs:
```diff
@@ -0,0 +1,179 @@
+
+            if (h2Node is null)
+            {
+                return new SearchResult(ExternalProviderType.KPopLyrics, ResponseStatusCode.NoDataFound);
```
Suggest adding logging to catch errors early and notice them. We expect the page to have lyrics by this point.
Let's add something like:
` _logger?.LogWarning($"KPopLyrics. Can't parse lyric from the page. Uri: {uri.ToString()}");`

skuill commented on LyricsScraperNET/Providers/KPopLyrics/KPopLyricsProvider.cs:
```diff
@@ -0,0 +1,179 @@
+
+            if (string.IsNullOrEmpty(rawHtmlLyrics))
+            {
+                return new SearchResult(ExternalProviderType.KPopLyrics, ResponseStatusCode.NoDataFound);
```
Suggest adding logging to catch errors early and notice them. We expect the page to have lyrics by this point.
Let's add something like:
` _logger?.LogWarning($"KPopLyrics. Can't parse lyric from the page. Uri: {uri.ToString()}");`

skuill reviewed (commented):
Thank you, @Lukeuke, for your effort and this PR! I’ve left a few comments and noticed the following improvements that need to be addressed:

1. **Add a new option to `appsettings.json`**: Please include the `KPopLyricsOptions` in the `appsettings.json` configuration of the `LyricsScraperNET.Client` console application.

2. **Handle missing lyrics case**: 
   If you run the console application with the following settings:
   ```csharp
   string artistToSearch = "Parkway Drive";
   string songToSearch = "Idols And Anchors";
   ```
   You’ll encounter the following error:
![image](https://github.com/user-attachments/assets/d4a5b819-3c51-484d-8e10-5a78445441fd)

   In this case, we’re requesting lyrics that don’t exist on the site, and this scenario should be handled and checked in the code. For example, the URI (`https://www.kpoplyrics.net/parkway-drive-idols-and-anchors-lyrics-english-romanized.html`) returns a "Page not found" (404). We need to validate this in the code:
   
![image](https://github.com/user-attachments/assets/ea39df8a-2adc-46ce-b555-890daf92c4ed)


3. **Incorrect URI generation for existing lyrics**:
   I tried the following case for an existing artist:
   ```csharp
   string artistToSearch = "CA$H KiD";
   string songToSearch = "CHANEL No.5";
   ```
   This resulted in an error:
![image](https://github.com/user-attachments/assets/c8374864-bbd8-42c0-b20b-a58c46009864)

   I noticed that we are attempting to fetch lyrics using an incorrect URI:
   `https://www.kpoplyrics.net/ca-h-kid-chanel-no-5-lyrics-english-romanized.html`
   The correct URI should be:
   `https://www.kpoplyrics.net/cah-kid-chanel-no5-lyrics-english-romanized.html`

4. **Add integration tests**: 
   In addition to the unit tests, I suggest adding integration tests for the cases identified during the testing above, as well as for the case already covered in the unit tests. Integration tests will help catch potential issues in the future, especially if the site updates and returns a new format for parsing.
5. Also, in the Readme.md, I suggest replacing, if you don't mind:
[kpoplyrics](https://www.kpoplyrics.net/) (**Coming soon** 🚧. [Issue #39](https://github.com/skuill/LyricsScraperNET/issues/39))
with: 
[kpoplyrics](https://www.kpoplyrics.net/) (added by [@Lukeuke](https://github.com/Lukeuke))

Thanks again! Feel free to reach out with any questions, happy to help!

Lukeuke committed:
KPopLyrics fixes:
- KPopLyricsProvider fixes
- Added integration tests for KPopLyricsProvider
- Removed "-lyrics-english-romanized.html" from KPopLyricsUriConverter
- Added KPopLyricsUriConverter Tests
- Fixed CreateCombinedUrlSlug method
- Added 1 more test case to GenerateCombinedUrlSlug_Tests

Lukeuke committed:
StringExtensions.cs whitespace fix

Lukeuke commented:
Hey, thanks for the review! 
@skuill I think I fixed everything you asked.

Changes:

- KPopLyricsProvider fixes
- Added integration tests for KPopLyricsProvider
- Removed "-lyrics-english-romanized.html" from KPopLyricsUriConverter
- Added KPopLyricsUriConverter Tests
- Fixed CreateCombinedUrlSlug method
- Added more test case to GenerateCombinedUrlSlug_Tests
- Whitespace fix
- Namespaces fix

Lukeuke committed:
Add KPopLyrics to appsettings.json

Lukeuke committed:
Update README.md (kpoplyrics)

Lukeuke committed:
Update README.md (kpoplyrics)

Lukeuke commented:
```csharp
string artistToSearch = "NINE.i (나인아이)";
string songToSearch = "Back to Christmas (크리스마스처럼)";
```
![image](https://github.com/user-attachments/assets/4feb44c5-9606-41aa-89fb-0955fac871ab)

Lukeuke committed:
Program.cs namespace fixes
KPopLyricsProvider.cs look for h2 containing "English"

skuill commented:
Hi @Lukeuke !
Thank you!
I've tried to run the console app and received the following result:
![image](https://github.com/user-attachments/assets/901bf427-ea90-470b-b091-f6313601e14d)

Also integration tests failed:
![image](https://github.com/user-attachments/assets/1a83a3ff-56b8-4e50-9d52-289beb675a1f)

Could you take a look please?

Lukeuke commented:
Hi! Thanks for bringing that up. I have had a look and I think i made i little mistake trying to fix code that is responsible for getting the lyrics from page.

Now its fixed and should be fine. I modified it a bit and also I added more integration tests to assure it's correct!

![image](https://github.com/user-attachments/assets/b3b5b319-123d-41b3-953b-addd291e2173)

What I've done:
I noticed that on the page are sometimes different types of lyrics:
- Official English Translation
- English Translation Lyrics
- Romanized
(Im trying to find the first if not, the second and only then Romanized version. After these 3 checks the code return NotFound.)

Also I added code that drops down the HTML tags that are left after the parsing. ex: <span> I've noticed that https://www.kpoplyrics.net/00s-psycho-lyrics-english-romanized.html - here are some spans with colored text so i need to also take that to an account. Finally Trim the lyrics from whitespaces

Lukeuke committed:
KPopLyricsProvider/Parser fixes & more integration tests

Lukeuke committed:
include Lyrics_Result_03.txt

skuill commented:
Hi @Lukeuke!

Thank you for your PR and for addressing the issue with retrieving lyrics from the different pages! 
Great job handling the different types of lyrics on the page and prioritizing them: official translations first, then English, and finally Romanized. The fallback to "NotFound" makes sense. I also like how you addressed leftover HTML tags (like in the https://www.kpoplyrics.net/00s-psycho-lyrics-english-romanized.html example) and added trimming for clean results.

I'll merge your PR and do a final review of the package before publishing it. I've also noted this as good case about adding the ability to search for lyrics in different languages. I'll update issue #1 once this is implemented.

Thanks again for your contribution! I'll let you know when the package is published.

Lukeuke commented:
Hi! @skuill Thank you for merging my PR!

Lukeuke commented:
Since know I know how its done i can try implementing one more provider in my spare time. I was thinking about #25 the html seems easy to scrape and the url slug should be also easy

skuill commented:
Hi @Lukeuke ! Sure, feel free to take any open issue or create a new one. 
The only problem with #25 is that there is some unique identifier in the URL, which must be taken from somewhere in order to correctly compose the request URL.
